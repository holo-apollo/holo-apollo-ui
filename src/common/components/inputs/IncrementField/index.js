// @flow
import * as React from 'react';
import autoBind from 'react-autobind';

import { Cont, StyledInput, Modifier } from './styled';

type Props = {
  initialValue: number,
  maxValue?: number,
  minValue?: number,
  step: number,
  onChange?: number => void,
  isPercent: boolean,
};

type State = {
  value: number,
};

const defaultProps = {
  initialValue: 0,
  step: 1,
  isPercent: false,
};

class IncrementField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      value: props.initialValue,
    };
    autoBind(this);
  }

  static defaultProps = defaultProps;

  decrement() {
    const { minValue, step, onChange } = this.props;
    const { value } = this.state;
    const nextValue = value - step;
    if (minValue === undefined || nextValue >= minValue) {
      this.setState({ value: nextValue });
      onChange && onChange(nextValue);
    }
  }

  increment() {
    const { maxValue, step, onChange } = this.props;
    const { value } = this.state;
    const nextValue = value + step;
    if (maxValue === undefined || nextValue <= maxValue) {
      this.setState({ value: nextValue });
      onChange && onChange(nextValue);
    }
  }

  render() {
    const value = this.props.isPercent
      ? `${this.state.value}%`
      : this.state.value.toString();
    return (
      <Cont>
        <Modifier onClick={this.decrement}>-</Modifier>
        <StyledInput
          type="text"
          value={value}
          disabled={true}
          size={value.length}
        />
        <Modifier onClick={this.increment}>+</Modifier>
      </Cont>
    );
  }
}

export default IncrementField;
