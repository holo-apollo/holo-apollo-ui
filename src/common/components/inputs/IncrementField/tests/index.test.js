import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import IncrementField from '../index';

describe('IncrementField', function() {
  it('should render normally', function() {
    const wrapper = mount(<IncrementField />);
    expect(wrapper.exists()).toBe(true);
  });

  it('should match snapshot', function() {
    const tree = renderer.create(<IncrementField />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
