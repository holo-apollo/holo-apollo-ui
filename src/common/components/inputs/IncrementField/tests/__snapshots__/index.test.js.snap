// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncrementField should match snapshot 1`] = `
<div
  className="styled__Cont-sc-1uczn2-0 kpQOVo"
>
  <div
    className="styled__Modifier-sc-1uczn2-2 cZesPT"
    onClick={[Function]}
  >
    -
  </div>
  <input
    className="styled__StyledInput-sc-1uczn2-1 bAFoTG"
    disabled={true}
    size={1}
    type="text"
    value="0"
  />
  <div
    className="styled__Modifier-sc-1uczn2-2 cZesPT"
    onClick={[Function]}
  >
    +
  </div>
</div>
`;
